lib/Service/RoomService.php

```diff
@@ -392,4 +409,101 @@ public function verifyPassword(Room $room, string $password): array {
+		$update->executeStatement();
+		$jobList = Server::get(IJobList::class);
+		if ($ttl > 0) {
+			$this->ttlSystemMessage($room, $participant, $ttl, 'ttl_enabled');
+			$jobList->add(ApplyTtl::class, ['room_id' => $room->getId()]);
+		} else {
+			$this->ttlSystemMessage($room, $participant, $ttl, 'ttl_disabled');
+			$jobList->remove(ApplyTtl::class, ['room_id' => $room->getId()]);
+		}
+
+		$this->dispatcher->dispatch(Room::EVENT_AFTER_SET_TIME_TO_LIVE, $event);
+	}
+
+	private function ttlSystemMessage(Room $room, Participant $participant, int $ttl, string $message): void {
+		$this->chatManager->addSystemMessage(
+			$room,
+			$participant->getAttendee()->getActorType(),
+			$participant->getAttendee()->getActorId(),
+			json_encode([
+				'message' => $message,
+				'parameters' => ['ttl' => $ttl]
+			]),
+			$this->timeFactory->getDateTime(),
+			false
+		);
+	}
+
+
+	public function deleteExpiredTtl(int $roomId, int $jobId): array {
+		$room = $this->manager->getRoomById($roomId);
+
+		$max = $this->getMaxDateTtl($room->getTimeToLive());
+		$min = $this->getMinDateTtl($jobId);
+
+		$ids = $this->commentsManager->getMessageIdsByRoomIdInDateInterval($roomId, $min, $max);
+		if (count($ids)) {
+			$this->reportDeletedMessagesIds($room, $ids);
+			$this->deleteMessagesByIds($ids);
+		}
+		return $ids;
+	}
+
+	private function getMaxDateTtl(int $ttl): \DateTime {
+		$max = $this->timeFactory->getDateTime();
+		return $max->sub(new DateInterval('PT' . $ttl . 'S'));
+	}
+
+	private function getMinDateTtl(int $jobId): \DateTime {
+		$query = $this->db->getQueryBuilder();
+		$query->select('last_checked')
+			->from('jobs')
+			->where(
+				$query->expr()->eq('id', $query->createNamedParameter($jobId, IQueryBuilder::PARAM_INT))
+			);
+		$result = $query->executeQuery();
+		$lastCheckedEpoch = $result->fetchOne();
+		$lastChechedDateTime = $this->timeFactory->getDateTime('@' . $lastCheckedEpoch);
+		return $lastChechedDateTime;
+	}
+
+	private function deleteMessagesByIds(array $ids): void {
+		$delete = $this->db->getQueryBuilder();
+		$delete->delete('comments')
+			->where(
+				$delete->expr()->orX(
+					$delete->expr()->in('id', $delete->createNamedParameter($ids, IQueryBuilder::PARAM_INT_ARRAY)),
+					$delete->expr()->in('parent_id', $delete->createNamedParameter($ids, IQueryBuilder::PARAM_INT_ARRAY)),
+					$delete->expr()->in('topmost_parent_id', $delete->createNamedParameter($ids, IQueryBuilder::PARAM_INT_ARRAY))
+				)
+			);
+		$delete->executeStatement();
+	}
+
+	private function reportDeletedMessagesIds(Room $chat, array $ids): void {
+		foreach ($ids as $id) {
+			$this->chatManager->addSystemMessage(
+				$chat,
+				'ttl_expired',
+				'ttl_expired',
+				json_encode(['message' => 'message_deleted', 'parameters' => ['message' => $id]]),
+				$this->timeFactory->getDateTime(),
+				false,
+				null,
+				$id,
+				true
+			);
+		}
+	}
```

Comment:
We shouldn't create a system message that a message has been deleted...

Reply:
How we will notify the clients to remove the deleted messages from message list? I did the same that when the user delete a message.

Reply:
They will have to respect the expiration timestamp we give when rendering the message.
The problem is this message ^ will again expire? :P 

But it seems you forgot about the change o approach we discussed last week or 2 weeks ago.
We need to bring a dedicated column to the oc_comments table with a PR in server.